tracer/tracer.go

```diff
@@ -723,8 +711,16 @@ func (t *Tracer) insertKernelFrames(trace *host.Trace, ustackLen uint32,
 		// Kernel frame PCs need to be adjusted by -1. This duplicates logic done in the trace
 		// converter. This should be fixed with PF-1042.
 		if foundSymbol && foundFileID {
-			t.reportFallbackKernelSymbol(fileID, symbol, trace.Frames[i].Lineno-1,
-				&kernelSymbolCacheHit, &kernelSymbolCacheMiss)
+			frameID := libpf.NewFrameID(fileID, trace.Frames[i].Lineno-1)
+			if t.reporter.FrameKnown(frameID) {
+				kernelSymbolCacheHit++
+			} else {
+				t.reporter.FrameMetadata(&reporter.FrameMetadataArgs{
+					FrameID:      frameID,
+					FunctionName: string(symbol),
+				})
+				kernelSymbolCacheMiss++
+			}
```

Comment:
Is the miss/hit statistics important here? The reporter API says calling `FrameKnown` is optional, and since all the data is here at hand, we could just omit it.

Alternatively we could move the `FrameKnown` check to be first in the `if`, and only if the frame is not known we do the `t.kernelSymbols.LookupByAddress` plus related debug logging. At least we'd get then rid of one search to make the extra `FrameKnown` call meaningful.

Reply:
> Is the miss/hit statistics important here?

I'd like to avoid mixing this unrelated decision into the PR and rather have a second PR for discussion.

> Alternatively we could move the FrameKnown check to be first in the if ...

The frame's fileID and address are retrieved by `t.kernelModules.LookupByAddress()` + ` t.moduleFileIDs[string(mod)]` lookup. How can this be short-circuited?

> The reporter API says calling FrameKnown is optional, and since all the data is here at hand, we could just omit it.

Yes we could, but that makes assumptions about the implementations of `FrameKnown` and `FrameMetadata`. I'd assume `FrameKnown` to be less resource intensive than `FrameMetadata`. Btw, looking into the code reveals that indeed `FrameMetadata` does at least one heap allocation while `FrameKnown` doesn't.

Reply:
> > Is the miss/hit statistics important here?
> 
> I'd like to avoid mixing this unrelated decision into the PR and rather have a second PR for discussion.

You are changing the API here and I believe the miss/hit statistics become meaningless because of this. I would like to address this now as it can simplify the PR.

> > Alternatively we could move the FrameKnown check to be first in the if ...
> 
> The frame's fileID and address are retrieved by `t.kernelModules.LookupByAddress()` + ` t.moduleFileIDs[string(mod)]` lookup. How can this be short-circuited?

I was trying to imply this:
``` diff
--- a/tracer/tracer.go
+++ b/tracer/tracer.go
@@ -683,8 +683,6 @@ func (t *Tracer) insertKernelFrames(trace *host.Trace, ustackLen uint32,
                //    LOAD segments. the address is relative to the .text section
                mod, addr, _ := t.kernelModules.LookupByAddress(
                        libpf.SymbolValue(kstackVal[i]))
-               symbol, offs, foundSymbol := t.kernelSymbols.LookupByAddress(
-                       libpf.SymbolValue(kstackVal[i]))
 
                fileID, foundFileID := t.moduleFileIDs[string(mod)]
 
@@ -692,8 +690,6 @@ func (t *Tracer) insertKernelFrames(trace *host.Trace, ustackLen uint32,
                        fileID = libpf.UnknownKernelFileID
                }
 
-               log.Debugf(" kstack[%d] = %v+%x (%v+%x)", i, string(mod), addr, symbol, offs)
-
                hostFileID := host.FileIDFromLibpf(fileID)
                t.processManager.FileIDMapper.Set(hostFileID, fileID)
 
@@ -710,17 +706,19 @@ func (t *Tracer) insertKernelFrames(trace *host.Trace, ustackLen uint32,
 
                // Kernel frame PCs need to be adjusted by -1. This duplicates logic done in the trace
                // converter. This should be fixed with PF-1042.
-               if foundSymbol && foundFileID {
-                       frameID := libpf.NewFrameID(fileID, trace.Frames[i].Lineno-1)
-                       if t.reporter.FrameKnown(frameID) {
-                               kernelSymbolCacheHit++
-                       } else {
+               frameID := libpf.NewFrameID(fileID, trace.Frames[i].Lineno-1)
+               if foundFileID && !t.reporter.FrameKnown(frameID) {
+                       symbol, _, foundSymbol := t.kernelSymbols.LookupByAddress(
+                               libpf.SymbolValue(kstackVal[i]))
+                       if foundSymbol {
                                t.reporter.FrameMetadata(&reporter.FrameMetadataArgs{
                                        FrameID:      frameID,
                                        FunctionName: string(symbol),
                                })
                                kernelSymbolCacheMiss++
                        }
+               } else {
+                       kernelSymbolCacheHit++
                }
        }
        t.fallbackSymbolMiss.Add(kernelSymbolCacheMiss)
```

> > The reporter API says calling FrameKnown is optional, and since all the data is here at hand, we could just omit it.
> 
> Yes we could, but that makes assumptions about the implementations of `FrameKnown` and `FrameMetadata`. I'd assume `FrameKnown` to be less resource intensive than `FrameMetadata`. Btw, looking into the code reveals that indeed `FrameMetadata` does at least one heap allocation while `FrameKnown` doesn't.

We have several implementations of reporter. Please do not make suggestions based on the one reporter in this repository.

The API documentatio for `FrameKnown` is clear that it is optional to be called. And in fact some of the interpreters already skip it.

The otlp reporter `FrameMetadata` should be fixed if it behaves badly when `FrameKnown` is not called.

Reply:
> We have several implementations of reporter. Please do not make suggestions based on the one reporter in this repository.

Sorry for the misunderstanding. That's exactly what I tried to say with "but that makes assumptions about the implementations of FrameKnown and FrameMetadata." =)

Applied your suggestion.

> I believe the miss/hit statistics become meaningless

I don't know if anybody makes use of these metrics currently, maybe not.
I'd suggest to remove them completely here and rather have metrics about caching in the reporter (in FrameMetadata).

> The otlp reporter FrameMetadata should be fixed if it behaves badly when FrameKnown is not called.

I'll have a look at this independently of this PR.

Reply:
Created #188 and #189 fir the two open points.